---
title: "R Notebook"
output: html_notebook
---

# Analysis of bipartite networks from SparXCC (Day and Night)

## Importing the edge lists for Day and Night


Correlations were obtained with all results from SparXCC having values above m (SparXCC permutation threshold).

Importing the bipartite networks from SparXCC:

```{r}
library(igraph)

# Reading the edge list with node names and correlations
day_edges <- read.csv("/media/rsantos/4TB_drive/Projects/UGA_RACS/CoNekT_Grasses_Microbiome/Transcriptome_Microbiome/SparXCC/PAG2025/SparXCC_output_day_pag25_edge_list.txt", sep="\t", header=TRUE)
night_edges <- read.csv("/media/rsantos/4TB_drive/Projects/UGA_RACS/CoNekT_Grasses_Microbiome/Transcriptome_Microbiome/SparXCC/PAG2025/SparXCC_output_night_pag25_edge_list.txt", sep="\t", header=TRUE)

# Assigning the node types
day_node_types <- c(
  setNames(rep(TRUE, length(unique(day_edges$OTU))), unique(day_edges$OTU)),  # OTU (TRUE)
  setNames(rep(FALSE, length(unique(day_edges$Gene))), unique(day_edges$Gene))  # Gene (FALSE)
)
night_node_types <- c(
  setNames(rep(TRUE, length(unique(night_edges$OTU))), unique(night_edges$OTU)),  # OTU (TRUE)
  setNames(rep(FALSE, length(unique(night_edges$Gene))), unique(night_edges$Gene))  # Gene (FALSE)
)

# Creating the graph objects (considering positive and negative correlations)
day_bipartite_graph <- graph_from_data_frame(d = day_edges, directed = FALSE)
night_bipartite_graph <- graph_from_data_frame(d = night_edges, directed = FALSE)

V(day_bipartite_graph)$type <- day_node_types[V(day_bipartite_graph)$name]
E(day_bipartite_graph)$cor <- day_edges$CorrelationCoefficient
print(summary(day_bipartite_graph))
print(data.frame(name = V(day_bipartite_graph)$name, type = V(bipartite_graph)$type))

day_node_colors <- ifelse(V(day_bipartite_graph)$type, "lightblue", "pink")

layout <- layout_with_fr(day_bipartite_graph)  # Force-directed layout
layout <- norm_coords(layout, ymin = -2, ymax = 2, xmin = -2, xmax = 2)  # Rescale coordinates for better spacing

plot(
  day_bipartite_graph,
  layout = layout,
  vertex.color = day_node_colors,
  vertex.label = V(day_bipartite_graph)$name,
  vertex.size = 10,
  vertex.label.cex = 0.1,
  edge.width = E(day_bipartite_graph)$cor  # Use weights to set edge widths
)
```

(This initial draft/plot does not show edges with negative values, since edge.width is set to the correlation values)

Interesting questions would be:

 * What are genes and OTUs that are shared, unique or different between day and night?
 * What are the edges that are shared, unique or different between day and night?
 
 
```{r}

```

## Analyzing networks

Converting correlations to absolute values:

```{r}
# Convert correlations to absolute values
day_edges$CorrelationCoefficient <- abs(day_edges$CorrelationCoefficient)
night_edges$CorrelationCoefficient <- abs(night_edges$CorrelationCoefficient)

# Creating the graph objects considering absolute values only (magnitude of associations are revelant)
day_bipartite_graph_abs <- graph_from_data_frame(d = day_edges, directed = FALSE)
night_bipartite_graph_abs <- graph_from_data_frame(d = night_edges, directed = FALSE)

# Node types were carried out previously

# Assigning the node types
V(day_bipartite_graph_abs)$type <- day_node_types[V(day_bipartite_graph_abs)$name]
E(day_bipartite_graph_abs)$cor <- day_edges$CorrelationCoefficient
V(night_bipartite_graph_abs)$type <- night_node_types[V(night_bipartite_graph_abs)$name]
E(night_bipartite_graph_abs)$cor <- night_edges$CorrelationCoefficient


```


Generating the biadjacency matrix and computing network modules:

```{r}
library(bipartite)

# bipartite uses a biadjacency matrix to represent the network
# Converting the graph to a biadjacency matrix using igraph function
day_biadjacency_abs <- as_biadjacency_matrix(day_bipartite_graph_abs,
                      types = V(day_bipartite_graph_abs)$type, # Node types (OTU and Gene)
                      attr="cor") # Correlations (absolute values)
night_biadjacency_abs <- as_biadjacency_matrix(night_bipartite_graph_abs,
                      types = V(night_bipartite_graph_abs)$type, # Node types (OTU and Gene)
                      attr="cor") # Correlations (absolute values)

# Compute network modules
day_modules_abs <- computeModules(day_biadjacency_abs,
                                  method="Beckett",
                                  forceLPA=FALSE)
night_modules_abs <- computeModules(night_biadjacency_abs,
                                  method="Beckett",
                                  forceLPA=FALSE)

node_information_day_modules_abs <- listModuleInformation(day_modules_abs)
node_information_night_modules_abs <- listModuleInformation(night_modules_abs)

length(node_information_day_modules_abs[[2]])
length(node_information_night_modules_abs[[2]])

# Create empty dataframe to store module information and fill it from the list
day_module_info_abs <- data.frame(matrix(ncol = 3, nrow = 0))
night_module_info_abs <- data.frame(matrix(ncol = 3, nrow = 0))
colnames(day_module_info_abs) <- c("Module", "ModuleSizeGene", "ModuleSizeOTU")
colnames(night_module_info_abs) <- c("Module", "ModuleSizeGene", "ModuleSizeOTU")

# Deep 2 modules
for (i in 1:length(node_information_day_modules_abs[[2]])) {
 day_module_info_abs <- rbind(day_module_info_abs, data.frame(Module = i,
                                                               ModuleSizeGene = length(node_information_day_modules_abs[[2]][[i]][[1]]),
                                                               ModuleSizeOTU = length(node_information_day_modules_abs[[2]][[i]][[2]])))
}

# Deep 2 modules
for (i in 1:length(node_information_night_modules_abs[[2]])) {
 night_module_info_abs <- rbind(night_module_info_abs, data.frame(Module = i,
                                                               ModuleSizeGene = length(node_information_night_modules_abs[[2]][[i]][[1]]),
                                                               ModuleSizeOTU = length(node_information_night_modules_abs[[2]][[i]][[2]])))
}

day_module_info_abs
night_module_info_abs

# Export list of genes for each module
# day
for (i in 1:length(node_information_day_modules_abs[[2]])) {
  write.table(node_information_day_modules_abs[[2]][[i]][[1]],
          paste("day_mod_", i, "_genes.txt", sep=""),
          quote=FALSE,
          sep="\t",
          col.names=FALSE,
          row.names=FALSE)
}
for (i in 1:length(node_information_day_modules_abs[[2]])) {
  write.table(node_information_day_modules_abs[[2]][[i]][[2]],
          paste("day_mod_", i, "_otus.txt", sep=""),
          quote=FALSE,
          sep="\t",
          col.names=FALSE,
          row.names=FALSE)
}

# night
?write.csv
for (i in 1:length(node_information_night_modules_abs[[2]])) {
  write.table(node_information_night_modules_abs[[2]][[i]][[1]],
          paste("night_mod_", i, "_genes.txt", sep=""),
          quote=FALSE,
          sep="\t",
          col.names=FALSE,
          row.names=FALSE)
}
for (i in 1:length(node_information_night_modules_abs[[2]])) {
  write.table(node_information_night_modules_abs[[2]][[i]][[2]],
          paste("night_mod_", i, "_otus.txt", sep=""),
          quote=FALSE,
          sep="\t",
          col.names=FALSE,
          row.names=FALSE)
}
```

```{r}
library(ggplot2)
library(reshape2)

day_module_info_abs
day_module_info_abs_melt <- melt(day_module_info_abs, id.vars='Module')

ggplot(day_module_info_abs_melt, aes(x=Module, y=value, fill=variable)) +
    geom_bar(stat='identity', position='dodge')
```

```{r}

night_module_info_abs
night_module_info_abs_melt <- melt(night_module_info_abs, id.vars='Module')
night_module_info_abs_melt

ggplot(night_module_info_abs_melt, aes(x=Module, y=value, fill=variable)) +
    geom_bar(stat='identity', position='dodge')
```

Interesting questions would be:

 * What modules are shared or unique between day and night? (regardless of correlation signs)


Obtaining metrics such as connectance, modularity and robustness.

```{r}
# Here is a list of metrics that can be computed with the bipartite library:
# Connectance
# Web asymmetry
# Links per species
# Number of compartments
# Compartment diversity
# Cluster coefficient
# Nestedness
# Weighted nestedness
# Weighted NODF (Almeida-Neto & Ulrich, 2011; method employed by Zhao et al 2024 - 10.1111/mec.17442)
# Interaction strength asymmetry
# Interaction asymmetry
# Specialization asymmetry
# Linkage density
# Fisher alpha, diversity index
# Alatalo interaction evenness
# Shannon diversity index
# Network-wide specialization index

networklevel_day_abs <- networklevel(day_biadjacency_abs,
             index="ALL",
             level="both",
             weighted=TRUE,
             ISAmethod="Bluethgen",
             extinctmethod="degree",
             nrep=100,
             CCfun=median,
             dist="horn",
             normalise=TRUE,
             empty.web=TRUE,
             intereven="sum",
             H2_integer=FALSE,
             fcweighted=TRUE,
             fcdist="euclidean",
             effective=FALSE)

networklevel_night_abs <- networklevel(night_biadjacency_abs,
             index="ALL",
             level="both",
             weighted=TRUE,
             ISAmethod="Bluethgen",
             extinctmethod="degree",
             nrep=100,
             CCfun=median,
             dist="horn",
             normalise=TRUE,
             empty.web=TRUE,
             intereven="sum",
             H2_integer=FALSE,
             fcweighted=TRUE,
             fcdist="euclidean",
             effective=FALSE)
```



