---
title: "Co-expression network analysis"
output: html_notebook
---

Following the [SimpleTidy GeneCoEx workflow](https://github.com/cxli233/SimpleTidy_GeneCoEx) created by Dr. Chenxin Li for the analysis of co-expression networks.

Importing the necessary libraries for the analysis:

```{r}
library(tidyverse)
#install.packages("tidyverse")
library(igraph)
library(ggraph)
#install.packages("ggraph")
library(readxl)
#install.packages("readxl")
library(patchwork)
#install.packages("patchwork")
library(RColorBrewer)
library(viridis)
```

Importing the TPM gene expression matrix. It was generated by running Salmon on the RNA-seq data from the Kremling et al. (2018) paper and using the index of Maize v5 as the reference transcriptome. The matrix was filtered by TPM and coefficient of variation (CV) values.

```{r}
Exp_table <- read_tsv("/home/rsantos/Repositories/maize_microbiome_transcriptomics/poster_presentation_notebooks/plant_center_retreat_2024/kremling_expression_v5_tpm_filtered_cv_filtered.tsv", col_types = cols())
Exp_table
```

Generating the Tidy data frame for the gene expression matrix:

```{r}
Exp_table_long <- Exp_table %>% 
  pivot_longer(cols = !Name, names_to = "SampleName", values_to = "TPM") %>% 
  mutate(logTPM = log10(tpm + 1))

head(Exp_table_long)
```

Generating a wide version again (with log TPM values):

```{r}
Exp_table_log_wide <- Exp_table_long %>% 
  select(Name, SampleName, logTPM) %>% 
  pivot_wider(names_from = SampleName, values_from = logTPM)

head(Exp_table_log_wide)
```

Importing metadata:

```{r}
sample_metadata <- read_tsv("/home/rsantos/Repositories/maize_microbiome_transcriptomics/poster_presentation_notebooks/pag_2025/sample_annotation.txt",
                            show_col_types = FALSE)
head(sample_metadata)
```

Generating a PCA:

```{r}
my_pca <- prcomp(t(Exp_table_log_wide[, -1]))
pc_importance <- as.data.frame(t(summary(my_pca)$importance))
head(pc_importance, 20)
```

```{r}
head(as.data.frame(my_pca$x[, 1:10]))
```

Adding metadata to the PCA matrix:

```{r}
PCA_coord <- my_pca$x[, 1:10] %>% 
  as.data.frame() %>% 
  mutate(SampleName = row.names(.)) %>% 
  full_join(sample_metadata %>% 
              select(Genotype, Subpopulation, DayPeriod, Tissue, SampleName), by = "SampleName")

head(PCA_coord)
```

```{r}
PCA_by_day_period <- PCA_coord %>% 
  ggplot(aes(x = PC1, y = PC2)) +
  geom_point(aes(fill = DayPeriod), color = "grey20", shape = 21, size = 3, alpha = 0.8) +
  scale_fill_manual(values = brewer.pal(n = 3, "Accent")) +
  labs(x = paste("PC1 (", pc_importance[1, 2] %>% signif(3)*100, "% of Variance)", sep = ""), 
       y = paste("PC2 (", pc_importance[2, 2] %>% signif(3)*100, "% of Variance)", "  ", sep = ""),
       fill = NULL) +  
  theme_bw() +
  theme(
    text = element_text(size= 14),
    axis.text = element_text(color = "black")
  )

PCA_by_day_period
```

```{r}
PCA_by_subpopulation <- PCA_coord %>% 
  ggplot(aes(x = PC1, y = PC2)) +
  geom_point(aes(fill = Subpopulation), color = "grey20", shape = 21, size = 3, alpha = 0.8) +
  labs(x = paste("PC1 (", pc_importance[1, 2] %>% signif(3)*100, "% of Variance)", sep = ""), 
       y = paste("PC2 (", pc_importance[2, 2] %>% signif(3)*100, "% of Variance)", "  ", sep = ""),
       fill = NULL) +  
  theme_bw() +
  theme(
    text = element_text(size= 14),
    axis.text = element_text(color = "black")
  )

PCA_by_subpopulation
```

Importing the correlation matrices. I (RACS) generated two different networks:

 * Day
 * Night

```{r}
day_cor_matrix <- as.data.frame(read_tsv("/home/rsantos/Repositories/maize_microbiome_transcriptomics/poster_presentation_notebooks/pag_2025/co_exp_day_pearson_correlation.tsv", col_types = cols()))
rownames(day_cor_matrix) <- day_cor_matrix[, 1]
day_cor_matrix <- day_cor_matrix[, -1]
night_cor_matrix <- as.data.frame(read_tsv("/home/rsantos/Repositories/maize_microbiome_transcriptomics/poster_presentation_notebooks/pag_2025/co_exp_night_pearson_correlation.tsv", col_types = cols()))
rownames(night_cor_matrix) <- night_cor_matrix[, 1]
night_cor_matrix <- night_cor_matrix[, -1]
```


```{r}


day_cor_matrix_tri <- day_cor_matrix
day_cor_matrix_tri[lower.tri(day_cor_matrix_tri)] <- NA
head(day_cor_matrix_tri)
```

For a initial edge selection, I (RACS) will use a correlation threshold of 0.7 (p-values and corrected p-values generated by corALS, so I will probably look at these values later):

```{r}
day_cor_matrix %>%
  as.data.frame()
```


Bait genes (we are probably going to use circadian genes, since we are comparing day and night periods):

```{r}

```

Major factors in the experiment:

```{r}
#Counting samples of different maize subpopulations
sample_metadata %>% 
  group_by(Subpopulation) %>% 
  count()
```

```{r}
#Counting samples of different day periods
sample_metadata %>% 
  group_by(DayPeriod) %>% 
  count()
```

