{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Running correlations bteween gene expression and OTU abundance\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filtering steps (low abundance and low expression, or low variance genes/OTUs) were described in another notebook.\n",
    "Here, I (RACS) use all filtered matrices for running correlations for each group of maize genotypes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Importing the within sample (RPKM, TPM, CPM) and between sample (TMM, UQ, CTF, CUF) normalization matrices for day and night:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Importing within sample normalized data after filtering\n",
    "#kremling_expression_v5_day_tpm_filtered_cv_filtered = pd.read_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_day_tpm_filtered_cv_filtered.tsv',\n",
    "kremling_expression_v5_day_tpm_filtered_cv_filtered = pd.read_csv('/home/rsantos/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_day_tpm_filtered_cv_filtered.tsv',\n",
    "                            sep='\\t')\n",
    "kremling_expression_v5_day_tpm_filtered_cv_filtered.set_index('Name', inplace=True)\n",
    "#kremling_expression_v5_night_tpm_filtered_cv_filtered = pd.read_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_night_tpm_filtered_cv_filtered.tsv',\n",
    "kremling_expression_v5_night_tpm_filtered_cv_filtered = pd.read_csv('/home/rsantos/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_night_tpm_filtered_cv_filtered.tsv',\n",
    "                            sep='\\t')\n",
    "kremling_expression_v5_night_tpm_filtered_cv_filtered.set_index('Name', inplace=True)\n",
    "#kremling_expression_v5_day_rpkm_filtered_cv_filtered = pd.read_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_day_rpkm_filtered_cv_filtered.tsv',\n",
    "kremling_expression_v5_day_rpkm_filtered_cv_filtered = pd.read_csv('/home/rsantos/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_day_rpkm_filtered_cv_filtered.tsv',\n",
    "                            sep='\\t')\n",
    "kremling_expression_v5_day_rpkm_filtered_cv_filtered.set_index('Name', inplace=True)\n",
    "#kremling_expression_v5_night_rpkm_filtered_cv_filtered = pd.read_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_night_rpkm_filtered_cv_filtered.tsv',\n",
    "kremling_expression_v5_night_rpkm_filtered_cv_filtered = pd.read_csv('/home/rsantos/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_night_rpkm_filtered_cv_filtered.tsv',\n",
    "                            sep='\\t')\n",
    "kremling_expression_v5_night_rpkm_filtered_cv_filtered.set_index('Name', inplace=True)\n",
    "#kremling_expression_v5_day_cpm_filtered_cv_filtered = pd.read_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_day_cpm_filtered_cv_filtered.tsv',\n",
    "kremling_expression_v5_day_cpm_filtered_cv_filtered = pd.read_csv('/home/rsantos/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_day_cpm_filtered_cv_filtered.tsv',\n",
    "                            sep='\\t')\n",
    "kremling_expression_v5_day_cpm_filtered_cv_filtered.set_index('Name', inplace=True)\n",
    "#kremling_expression_v5_night_cpm_filtered_cv_filtered = pd.read_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_night_cpm_filtered_cv_filtered.tsv',\n",
    "kremling_expression_v5_night_cpm_filtered_cv_filtered = pd.read_csv('/home/rsantos/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_night_cpm_filtered_cv_filtered.tsv',\n",
    "                            sep='\\t')\n",
    "kremling_expression_v5_night_cpm_filtered_cv_filtered.set_index('Name', inplace=True)\n",
    "\n",
    "# Importing between sample normalized data after filtering\n",
    "#kremling_expression_v5_day_tmm_filtered_cv_filtered = pd.read_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_day_tmm_filtered_cv_filtered.tsv',\n",
    "kremling_expression_v5_day_tmm_filtered_cv_filtered = pd.read_csv('/home/rsantos/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_day_tmm_filtered_cv_filtered.tsv',\n",
    "                            sep='\\t')\n",
    "kremling_expression_v5_day_tmm_filtered_cv_filtered.set_index('Name', inplace=True)\n",
    "#kremling_expression_v5_night_tmm_filtered_cv_filtered = pd.read_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_night_tmm_filtered_cv_filtered.tsv',\n",
    "kremling_expression_v5_night_tmm_filtered_cv_filtered = pd.read_csv('/home/rsantos/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_night_tmm_filtered_cv_filtered.tsv',\n",
    "                            sep='\\t')\n",
    "kremling_expression_v5_night_tmm_filtered_cv_filtered.set_index('Name', inplace=True)\n",
    "kremling_expression_v5_day_uq_filtered_cv_filtered = pd.read_csv('/home/rsantos/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_day_uq_filtered_cv_filtered.tsv',\n",
    "#kremling_expression_v5_day_uq_filtered_cv_filtered = pd.read_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_day_uq_filtered_cv_filtered.tsv',\n",
    "                            sep='\\t')\n",
    "kremling_expression_v5_day_uq_filtered_cv_filtered.set_index('Name', inplace=True)\n",
    "#kremling_expression_v5_night_uq_filtered_cv_filtered = pd.read_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_night_uq_filtered_cv_filtered.tsv',\n",
    "kremling_expression_v5_night_uq_filtered_cv_filtered = pd.read_csv('/home/rsantos/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_night_uq_filtered_cv_filtered.tsv',\n",
    "                            sep='\\t')\n",
    "kremling_expression_v5_night_uq_filtered_cv_filtered.set_index('Name', inplace=True)\n",
    "#kremling_expression_v5_day_ctf_filtered_cv_filtered = pd.read_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_day_ctf_filtered_cv_filtered.tsv',\n",
    "kremling_expression_v5_day_ctf_filtered_cv_filtered = pd.read_csv('/home/rsantos/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_day_ctf_filtered_cv_filtered.tsv',\n",
    "                            sep='\\t')\n",
    "kremling_expression_v5_day_ctf_filtered_cv_filtered.set_index('Name', inplace=True)\n",
    "#kremling_expression_v5_night_ctf_filtered_cv_filtered = pd.read_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_night_ctf_filtered_cv_filtered.tsv',\n",
    "kremling_expression_v5_night_ctf_filtered_cv_filtered = pd.read_csv('/home/rsantos/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_night_ctf_filtered_cv_filtered.tsv',\n",
    "                            sep='\\t')\n",
    "kremling_expression_v5_night_ctf_filtered_cv_filtered.set_index('Name', inplace=True)\n",
    "#kremling_expression_v5_day_cuf_filtered_cv_filtered = pd.read_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_day_cuf_filtered_cv_filtered.tsv',\n",
    "kremling_expression_v5_day_cuf_filtered_cv_filtered = pd.read_csv('/home/rsantos/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_day_cuf_filtered_cv_filtered.tsv',\n",
    "                            sep='\\t')\n",
    "kremling_expression_v5_day_cuf_filtered_cv_filtered.set_index('Name', inplace=True)\n",
    "#kremling_expression_v5_night_cuf_filtered_cv_filtered = pd.read_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_night_cuf_filtered_cv_filtered.tsv',\n",
    "kremling_expression_v5_night_cuf_filtered_cv_filtered = pd.read_csv('/home/rsantos/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/quantification/kremling_expression_v5_night_cuf_filtered_cv_filtered.tsv',\n",
    "                            sep='\\t')\n",
    "kremling_expression_v5_night_cuf_filtered_cv_filtered.set_index('Name', inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Importing OTU (CPM) matrices for day and night:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importing OTU data after filtering\n",
    "#otu_table_merged_day_cpm_filtered_cv_filtered = pd.read_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/16S_wallace2018/combine_day_night_samples/otu_table_merged_day_cpm_filtered_cv_filtered.tsv',\n",
    "otu_table_merged_day_cpm_filtered_cv_filtered = pd.read_csv('/home/rsantos/Repositories/maize_microbiome_transcriptomics/16S_wallace2018/combine_day_night_samples/otu_table_merged_day_cpm_filtered_cv_filtered.tsv',\n",
    "                            sep='\\t')\n",
    "otu_table_merged_day_cpm_filtered_cv_filtered.set_index('OTU ID', inplace=True)\n",
    "#otu_table_merged_night_cpm_filtered_cv_filtered = pd.read_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/16S_wallace2018/combine_day_night_samples/otu_table_merged_night_cpm_filtered_cv_filtered.tsv',\n",
    "otu_table_merged_night_cpm_filtered_cv_filtered = pd.read_csv('/home/rsantos/Repositories/maize_microbiome_transcriptomics/16S_wallace2018/combine_day_night_samples/otu_table_merged_night_cpm_filtered_cv_filtered.tsv',\n",
    "                            sep='\\t')\n",
    "otu_table_merged_night_cpm_filtered_cv_filtered.set_index('OTU ID', inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Importing groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Accesion N</th>\n",
       "      <th>N GBS samples</th>\n",
       "      <th>N Plants</th>\n",
       "      <th>Avg. IBS</th>\n",
       "      <th>% missing</th>\n",
       "      <th>Breeding program</th>\n",
       "      <th>Pop structure</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Plot_Day</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14A0309_26</th>\n",
       "      <td>33-16</td>\n",
       "      <td>Ames26771</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.34</td>\n",
       "      <td>Other</td>\n",
       "      <td>unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14A0233_8</th>\n",
       "      <td>38-11</td>\n",
       "      <td>Ames26604</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.992</td>\n",
       "      <td>0.21</td>\n",
       "      <td>Other</td>\n",
       "      <td>unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14A0079_26</th>\n",
       "      <td>4226</td>\n",
       "      <td>NSL30904</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.29</td>\n",
       "      <td>Other</td>\n",
       "      <td>unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14A0311_8</th>\n",
       "      <td>4722</td>\n",
       "      <td>PI587130</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.39</td>\n",
       "      <td>Other</td>\n",
       "      <td>unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14A0021_8</th>\n",
       "      <td>A188</td>\n",
       "      <td>Ames22443</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0.29</td>\n",
       "      <td>Minnesota</td>\n",
       "      <td>unclassified</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Genotype Accesion N  N GBS samples  N Plants  Avg. IBS  % missing  \\\n",
       "Plot_Day                                                                       \n",
       "14A0309_26    33-16  Ames26771            4.0       3.0     0.998       0.34   \n",
       "14A0233_8     38-11  Ames26604            6.0       5.0     0.992       0.21   \n",
       "14A0079_26     4226   NSL30904            4.0       3.0     0.998       0.29   \n",
       "14A0311_8      4722   PI587130            3.0       2.0     0.998       0.39   \n",
       "14A0021_8      A188  Ames22443            8.0       7.0     0.996       0.29   \n",
       "\n",
       "           Breeding program Pop structure  \n",
       "Plot_Day                                   \n",
       "14A0309_26            Other  unclassified  \n",
       "14A0233_8             Other  unclassified  \n",
       "14A0079_26            Other  unclassified  \n",
       "14A0311_8             Other  unclassified  \n",
       "14A0021_8         Minnesota  unclassified  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#wallace_et_al_2018_group_assignments_romay2013 = pd.read_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/genotype_information/wallace_et_al_2018_group_assignments_romay2013.tsv',\n",
    "wallace_et_al_2018_group_assignments_romay2013 = pd.read_csv('/home/rsantos/Repositories/maize_microbiome_transcriptomics/genotype_information/wallace_et_al_2018_group_assignments_romay2013.tsv',\n",
    "                                                             sep='\\t')\n",
    "wallace_et_al_2018_group_assignments_romay2013.set_index('Plot_Day', inplace=True)\n",
    "wallace_et_al_2018_group_assignments_romay2013.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running correlations\n",
    "\n",
    "I (RACS) will start by analyzing correlations in three largest groups:\n",
    "* stiff stalk (39)\n",
    "* tropical (38)\n",
    "* non-stiff stalk (27)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/rsantos/Repositories/maize_microbiome_transcriptomics/.venv/lib/python3.10/site-packages/corals/threads.py:48: UserWarning: This function should be called before `numpy` or similar modules are imported.\n",
      "  warnings.warn(\"This function should be called before `numpy` or similar modules are imported.\")\n"
     ]
    }
   ],
   "source": [
    "from scipy import stats\n",
    "from corals.threads import set_threads_for_external_libraries\n",
    "set_threads_for_external_libraries(n_threads=1)\n",
    "import numpy as np\n",
    "from corals.correlation.full.default import cor_full"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### OTU (CPM) vs Gene (CPM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/rsantos/Repositories/maize_microbiome_transcriptomics/.venv/lib/python3.10/site-packages/corals/correlation/utils.py:227: RuntimeWarning: divide by zero encountered in divide\n",
      "  ts = rf * rf * (df / (1 - rf * rf))\n",
      "/home/rsantos/Repositories/maize_microbiome_transcriptomics/.venv/lib/python3.10/site-packages/corals/correlation/utils.py:227: RuntimeWarning: divide by zero encountered in divide\n",
      "  ts = rf * rf * (df / (1 - rf * rf))\n",
      "/home/rsantos/Repositories/maize_microbiome_transcriptomics/.venv/lib/python3.10/site-packages/corals/correlation/utils.py:227: RuntimeWarning: divide by zero encountered in divide\n",
      "  ts = rf * rf * (df / (1 - rf * rf))\n",
      "/home/rsantos/Repositories/maize_microbiome_transcriptomics/.venv/lib/python3.10/site-packages/corals/correlation/utils.py:227: RuntimeWarning: divide by zero encountered in divide\n",
      "  ts = rf * rf * (df / (1 - rf * rf))\n",
      "/home/rsantos/Repositories/maize_microbiome_transcriptomics/.venv/lib/python3.10/site-packages/corals/correlation/utils.py:227: RuntimeWarning: divide by zero encountered in divide\n",
      "  ts = rf * rf * (df / (1 - rf * rf))\n",
      "/home/rsantos/Repositories/maize_microbiome_transcriptomics/.venv/lib/python3.10/site-packages/corals/correlation/utils.py:227: RuntimeWarning: divide by zero encountered in divide\n",
      "  ts = rf * rf * (df / (1 - rf * rf))\n"
     ]
    }
   ],
   "source": [
    "concat_df_night = pd.concat([kremling_expression_v5_night_cpm_filtered_cv_filtered, otu_table_merged_night_cpm_filtered_cv_filtered], axis=0)\n",
    "concat_df_day = pd.concat([kremling_expression_v5_day_cpm_filtered_cv_filtered, otu_table_merged_day_cpm_filtered_cv_filtered], axis=0)\n",
    "\n",
    "concatenated_transposed_day = concat_df_day.transpose()\n",
    "concatenated_transposed_night = concat_df_night.transpose()\n",
    "\n",
    "shape_row_day = kremling_expression_v5_day_cpm_filtered_cv_filtered.shape[0]\n",
    "shape_row_night = kremling_expression_v5_night_cpm_filtered_cv_filtered.shape[0]\n",
    "\n",
    "# Selecting runs of different subpopulations (stiff stalk, non-stiff stalk and tropical) to filter dataframes\n",
    "stiff_stalk_runs_day = list(set(wallace_et_al_2018_group_assignments_romay2013[wallace_et_al_2018_group_assignments_romay2013['Pop structure']=='stiff stalk'].index).intersection(set(concatenated_transposed_day.index)))\n",
    "stiff_stalk_runs_night = list(set(wallace_et_al_2018_group_assignments_romay2013[wallace_et_al_2018_group_assignments_romay2013['Pop structure']=='stiff stalk'].index).intersection(set(concatenated_transposed_night.index)))\n",
    "non_stiff_stalk_runs_day = list(set(wallace_et_al_2018_group_assignments_romay2013[wallace_et_al_2018_group_assignments_romay2013['Pop structure']=='non-stiff stalk'].index).intersection(set(concatenated_transposed_day.index)))\n",
    "non_stiff_stalk_runs_night = list(set(wallace_et_al_2018_group_assignments_romay2013[wallace_et_al_2018_group_assignments_romay2013['Pop structure']=='non-stiff stalk'].index).intersection(set(concatenated_transposed_night.index)))\n",
    "tropical_runs_day = list(set(wallace_et_al_2018_group_assignments_romay2013[wallace_et_al_2018_group_assignments_romay2013['Pop structure']=='tropical'].index).intersection(set(concatenated_transposed_day.index)))\n",
    "tropical_runs_night = list(set(wallace_et_al_2018_group_assignments_romay2013[wallace_et_al_2018_group_assignments_romay2013['Pop structure']=='tropical'].index).intersection(set(concatenated_transposed_night.index)))\n",
    "\n",
    "# Calculating the correlations for each subpopulation\n",
    "stiff_stalk_cor_values_day = cor_full(concatenated_transposed_day.loc[stiff_stalk_runs_day])\n",
    "stiff_stalk_cor_values_night = cor_full(concatenated_transposed_night.loc[stiff_stalk_runs_night])\n",
    "non_stiff_stalk_cor_values_day = cor_full(concatenated_transposed_day.loc[non_stiff_stalk_runs_day])\n",
    "non_stiff_stalk_cor_values_night = cor_full(concatenated_transposed_night.loc[non_stiff_stalk_runs_night])\n",
    "tropical_cor_values_day = cor_full(concatenated_transposed_day.loc[tropical_runs_day])\n",
    "tropical_cor_values_night = cor_full(concatenated_transposed_night.loc[tropical_runs_night])\n",
    "\n",
    "# Calculating pvalues and corrected pvalues\n",
    "from corals.correlation.utils import derive_pvalues, multiple_test_correction\n",
    "\n",
    "#stiff_stalk_cor_values_day = cor_full(concatenated_transposed_day.loc[stiff_stalk_runs_day])\n",
    "n_samples = concatenated_transposed_day.loc[stiff_stalk_runs_day].shape[0]\n",
    "n_features = concatenated_transposed_day.loc[stiff_stalk_runs_day].shape[1]\n",
    "stiff_stalk_day_pvalues = derive_pvalues(stiff_stalk_cor_values_day, n_samples)\n",
    "stiff_stalk_day_pvalues_corrected = multiple_test_correction(stiff_stalk_day_pvalues, n_features, method=\"bonferroni\")\n",
    "\n",
    "#stiff_stalk_cor_values_night = cor_full(concatenated_transposed_night.loc[stiff_stalk_runs_night])\n",
    "n_samples = concatenated_transposed_night.loc[stiff_stalk_runs_night].shape[0]\n",
    "n_features = concatenated_transposed_night.loc[stiff_stalk_runs_night].shape[1]\n",
    "stiff_stalk_night_pvalues = derive_pvalues(stiff_stalk_cor_values_night, n_samples)\n",
    "stiff_stalk_night_pvalues_corrected = multiple_test_correction(stiff_stalk_night_pvalues, n_features, method=\"bonferroni\")\n",
    "\n",
    "#non_stiff_stalk_cor_values_day = cor_full(concatenated_transposed_day.loc[non_stiff_stalk_runs_day])\n",
    "n_samples = concatenated_transposed_day.loc[non_stiff_stalk_runs_day].shape[0]\n",
    "n_features = concatenated_transposed_day.loc[non_stiff_stalk_runs_day].shape[1]\n",
    "non_stiff_stalk_day_pvalues = derive_pvalues(non_stiff_stalk_cor_values_day, n_samples)\n",
    "non_stiff_stalk_day_pvalues_corrected = multiple_test_correction(non_stiff_stalk_day_pvalues, n_features, method=\"bonferroni\")\n",
    "\n",
    "#non_stiff_stalk_cor_values_night = cor_full(concatenated_transposed_night.loc[non_stiff_stalk_runs_night])\n",
    "n_samples = concatenated_transposed_night.loc[non_stiff_stalk_runs_night].shape[0]\n",
    "n_features = concatenated_transposed_night.loc[non_stiff_stalk_runs_night].shape[1]\n",
    "non_stiff_stalk_night_pvalues = derive_pvalues(non_stiff_stalk_cor_values_night, n_samples)\n",
    "non_stiff_stalk_night_pvalues_corrected = multiple_test_correction(non_stiff_stalk_night_pvalues, n_features, method=\"bonferroni\")\n",
    "\n",
    "#tropical_cor_values_day = cor_full(concatenated_transposed_day.loc[tropical_runs_day])\n",
    "n_samples = concatenated_transposed_day.loc[tropical_runs_day].shape[0]\n",
    "n_features = concatenated_transposed_day.loc[tropical_runs_day].shape[1]\n",
    "tropical_day_pvalues = derive_pvalues(tropical_cor_values_day, n_samples)\n",
    "tropical_day_pvalues_corrected = multiple_test_correction(tropical_day_pvalues, n_features, method=\"bonferroni\")\n",
    "\n",
    "#tropical_cor_values_night = cor_full(concatenated_transposed_night.loc[tropical_runs_night])\n",
    "n_samples = concatenated_transposed_night.loc[tropical_runs_night].shape[0]\n",
    "n_features = concatenated_transposed_night.loc[tropical_runs_night].shape[1]\n",
    "tropical_night_pvalues = derive_pvalues(tropical_cor_values_night, n_samples)\n",
    "tropical_night_pvalues_corrected = multiple_test_correction(tropical_night_pvalues, n_features, method=\"bonferroni\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9557850\n",
      "44394\n"
     ]
    }
   ],
   "source": [
    "# Selecting the pairs that have a correlation higher than threshold\n",
    "cor_threshold = 0.6\n",
    "cor_pvalue_threshold = 0.05\n",
    "\n",
    "stiff_stalk_true_positions_day_cor = np.where(stiff_stalk_cor_values_day > cor_threshold)\n",
    "stiff_stalk_true_positions_day_pval = np.where(stiff_stalk_day_pvalues_corrected < cor_pvalue_threshold)\n",
    "\n",
    "cor_tuples = []\n",
    "pval_tuples = []\n",
    "\n",
    "for i in range(np.size(stiff_stalk_true_positions_day_cor, 1)):\n",
    "    cor_tuples.append((stiff_stalk_true_positions_day_cor[0][i],\n",
    "    stiff_stalk_true_positions_day_cor[1][i]))\n",
    "\n",
    "for i in range(np.size(stiff_stalk_true_positions_day_pval, 1)):\n",
    "    pval_tuples.append((stiff_stalk_true_positions_day_pval[0][i],\n",
    "    stiff_stalk_true_positions_day_pval[1][i]))\n",
    "\n",
    "cor_tuples_set = set(cor_tuples)\n",
    "pval_tuples_set = set(pval_tuples)\n",
    "\n",
    "print(len(cor_tuples_set))\n",
    "print(len(pval_tuples_set))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "stiff_stalk_pairs_day_genecpm_otucpm = []\n",
    "stiff_stalk_pairs_night_genecpm_otucpm = []\n",
    "non_stiff_stalk_pairs_day_genecpm_otucpm = []\n",
    "non_stiff_stalk_pairs_night_genecpm_otucpm = []\n",
    "tropical_pairs_day_genecpm_otucpm = []\n",
    "tropical_pairs_night_genecpm_otucpm = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/rsantos/Repositories/maize_microbiome_transcriptomics/.venv/lib/python3.10/site-packages/corals/correlation/utils.py:227: RuntimeWarning: divide by zero encountered in divide\n",
      "  ts = rf * rf * (df / (1 - rf * rf))\n",
      "/home/rsantos/Repositories/maize_microbiome_transcriptomics/.venv/lib/python3.10/site-packages/corals/correlation/utils.py:227: RuntimeWarning: divide by zero encountered in divide\n",
      "  ts = rf * rf * (df / (1 - rf * rf))\n",
      "/home/rsantos/Repositories/maize_microbiome_transcriptomics/.venv/lib/python3.10/site-packages/corals/correlation/utils.py:227: RuntimeWarning: divide by zero encountered in divide\n",
      "  ts = rf * rf * (df / (1 - rf * rf))\n",
      "/home/rsantos/Repositories/maize_microbiome_transcriptomics/.venv/lib/python3.10/site-packages/corals/correlation/utils.py:227: RuntimeWarning: divide by zero encountered in divide\n",
      "  ts = rf * rf * (df / (1 - rf * rf))\n",
      "/home/rsantos/Repositories/maize_microbiome_transcriptomics/.venv/lib/python3.10/site-packages/corals/correlation/utils.py:227: RuntimeWarning: divide by zero encountered in divide\n",
      "  ts = rf * rf * (df / (1 - rf * rf))\n",
      "/home/rsantos/Repositories/maize_microbiome_transcriptomics/.venv/lib/python3.10/site-packages/corals/correlation/utils.py:227: RuntimeWarning: divide by zero encountered in divide\n",
      "  ts = rf * rf * (df / (1 - rf * rf))\n"
     ]
    }
   ],
   "source": [
    "# Storing the pairs that have a correlation higher than threshold\n",
    "for i in range(len(stiff_stalk_true_positions_day[0])):\n",
    "    if (stiff_stalk_true_positions_day[1][i] > (shape_row_day - 1)) and (stiff_stalk_true_positions_day[0][i] < shape_row_day):\n",
    "        stiff_stalk_pairs_day_genecpm_otucpm.append((str(stiff_stalk_cor_values_day.columns[stiff_stalk_true_positions_day[1][i]]),\n",
    "              str(stiff_stalk_cor_values_day.index[stiff_stalk_true_positions_day[0][i]])))\n",
    "\n",
    "for i in range(len(stiff_stalk_true_positions_night[0])):\n",
    "    if (stiff_stalk_true_positions_night[1][i] > (shape_row_night - 1)) and (stiff_stalk_true_positions_night[0][i] < shape_row_night):\n",
    "        stiff_stalk_pairs_night_genecpm_otucpm.append((str(stiff_stalk_cor_values_night.columns[stiff_stalk_true_positions_night[1][i]]),\n",
    "              str(stiff_stalk_cor_values_night.index[stiff_stalk_true_positions_night[0][i]])))\n",
    "        \n",
    "for i in range(len(non_stiff_stalk_true_positions_day[0])):\n",
    "    if (non_stiff_stalk_true_positions_day[1][i] > (shape_row_day - 1)) and (non_stiff_stalk_true_positions_day[0][i] < shape_row_day):\n",
    "        non_stiff_stalk_pairs_day_genecpm_otucpm.append((str(non_stiff_stalk_cor_values_day.columns[non_stiff_stalk_true_positions_day[1][i]]),\n",
    "              str(non_stiff_stalk_cor_values_day.index[non_stiff_stalk_true_positions_day[0][i]])))\n",
    "\n",
    "for i in range(len(non_stiff_stalk_true_positions_night[0])):\n",
    "    if (non_stiff_stalk_true_positions_night[1][i] > (shape_row_night - 1)) and (non_stiff_stalk_true_positions_night[0][i] < shape_row_night):\n",
    "        non_stiff_stalk_pairs_night_genecpm_otucpm.append((str(non_stiff_stalk_cor_values_night.columns[non_stiff_stalk_true_positions_night[1][i]]),\n",
    "              str(non_stiff_stalk_cor_values_night.index[non_stiff_stalk_true_positions_night[0][i]])))\n",
    "\n",
    "for i in range(len(tropical_true_positions_day[0])):\n",
    "    if (tropical_true_positions_day[1][i] > (shape_row_day - 1)) and (tropical_true_positions_day[0][i] < shape_row_day):\n",
    "        tropical_pairs_day_genecpm_otucpm.append((str(tropical_cor_values_day.columns[tropical_true_positions_day[1][i]]),\n",
    "              str(tropical_cor_values_day.index[tropical_true_positions_day[0][i]])))\n",
    "\n",
    "for i in range(len(tropical_true_positions_night[0])):\n",
    "    if (tropical_true_positions_night[1][i] > (shape_row_night - 1)) and (tropical_true_positions_night[0][i] < shape_row_night):\n",
    "        tropical_pairs_night_genecpm_otucpm.append((str(tropical_cor_values_night.columns[tropical_true_positions_night[1][i]]),\n",
    "              str(tropical_cor_values_night.index[tropical_true_positions_night[0][i]])))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Method using scipy\n",
    "\n",
    "Method from stats module of the Scipy library is much slower:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rng = np.random.default_rng()\n",
    "method = stats.PermutationMethod(n_resamples=10000, random_state=rng)\n",
    "\n",
    "significance_stiff_stalk_day = []\n",
    "significance_stiff_stalk_night = []\n",
    "significance_non_stiff_stalk_day = []\n",
    "significance_non_stiff_stalk_night = []\n",
    "significance_tropical_day = []\n",
    "significance_tropical_night = []\n",
    "\n",
    "for pair in stiff_stalk_pairs_day_genecpm_otucpm:\n",
    "    stat_res = stats.pearsonr(concatenated_transposed_day.loc[stiff_stalk_runs_day][pair[0]],\n",
    "               concatenated_transposed_day.loc[stiff_stalk_runs_day][pair[1]],\n",
    "               method=method)\n",
    "    significance_stiff_stalk_day.append({'otu': pair[0],\n",
    "                                         'gene': pair[1],\n",
    "                                         'pvalue': stat_res.pvalue,\n",
    "                                         'statistic': stat_res.statistic})\n",
    "\n",
    "for pair in stiff_stalk_pairs_night_genecpm_otucpm:\n",
    "    stat_res = stats.pearsonr(concatenated_transposed_night.loc[stiff_stalk_runs_night][pair[0]],\n",
    "               concatenated_transposed_night.loc[stiff_stalk_runs_night][pair[1]],\n",
    "               method=method)\n",
    "    significance_stiff_stalk_night.append({'otu': pair[0],\n",
    "                                         'gene': pair[1],\n",
    "                                         'pvalue': stat_res.pvalue,\n",
    "                                         'statistic': stat_res.statistic})\n",
    "\n",
    "for pair in non_stiff_stalk_pairs_day_genecpm_otucpm:\n",
    "    stat_res = stats.pearsonr(concatenated_transposed_day.loc[non_stiff_stalk_runs_day][pair[0]],\n",
    "               concatenated_transposed_day.loc[non_stiff_stalk_runs_day][pair[1]],\n",
    "               method=method)\n",
    "    significance_non_stiff_stalk_day.append({'otu': pair[0],\n",
    "                                         'gene': pair[1],\n",
    "                                         'pvalue': stat_res.pvalue,\n",
    "                                         'statistic': stat_res.statistic})\n",
    "\n",
    "for pair in non_stiff_stalk_pairs_night_genecpm_otucpm:\n",
    "    stat_res = stats.pearsonr(concatenated_transposed_night.loc[non_stiff_stalk_runs_night][pair[0]],\n",
    "               concatenated_transposed_night.loc[non_stiff_stalk_runs_night][pair[1]],\n",
    "               method=method)\n",
    "    significance_non_stiff_stalk_night.append({'otu': pair[0],\n",
    "                                         'gene': pair[1],\n",
    "                                         'pvalue': stat_res.pvalue,\n",
    "                                         'statistic': stat_res.statistic})\n",
    "\n",
    "for pair in tropical_pairs_day_genecpm_otucpm:\n",
    "    stat_res = stats.pearsonr(concatenated_transposed_day.loc[tropical_runs_day][pair[0]],\n",
    "               concatenated_transposed_day.loc[tropical_runs_day][pair[1]],\n",
    "               method=method)\n",
    "    significance_tropical_day.append({'otu': pair[0],\n",
    "                                         'gene': pair[1],\n",
    "                                         'pvalue': stat_res.pvalue,\n",
    "                                         'statistic': stat_res.statistic})\n",
    "\n",
    "for pair in tropical_pairs_night_genecpm_otucpm:\n",
    "    stat_res = stats.pearsonr(concatenated_transposed_night.loc[tropical_runs_night][pair[0]],\n",
    "               concatenated_transposed_night.loc[tropical_runs_night][pair[1]],\n",
    "               method=method)\n",
    "    significance_tropical_night.append({'otu': pair[0],\n",
    "                                         'gene': pair[1],\n",
    "                                         'pvalue': stat_res.pvalue,\n",
    "                                         'statistic': stat_res.statistic})"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
