{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generation of initial matrices and summary of methods\n",
    "\n",
    "Several methods for normalization and transformation of RNAseq will be tested from [Johnson and Krishnan 2022](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02568-9). A figure describing their analysis pipeline is [here](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02568-9/figures/1).\n",
    "\n",
    "\n",
    "## Filtering lowly expressed genes and lowly abundant OTUs\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## Normalization methods for RNAseq and metataxonomics\n",
    "\n",
    "\n",
    "| Data type      | Normalization         | Condition  |\n",
    "|----------------|-----------------------|------------|\n",
    "| Metataxonomics | Raw counts            | Joined D and N |\n",
    "| Metataxonomics | CPM                   | Joined D and N |\n",
    "| Metataxonomics | Relative abundance    | Joined D and N |\n",
    "| Transcriptomics | Estimated counts     | D |\n",
    "| Transcriptomics | CPM                  | D |\n",
    "| Transcriptomics | TPM                  | D |\n",
    "| Transcriptomics | TMM                  | D |\n",
    "| Transcriptomics | RPKM                 | D |\n",
    "| Transcriptomics | UQ                   | D |\n",
    "| Transcriptomics | CTF                  | D |\n",
    "| Transcriptomics | CUF                  | D |\n",
    "| Transcriptomics | Others (UPDATE THIS) | D |\n",
    "| Transcriptomics | Estimated counts     | N |\n",
    "| Transcriptomics | CPM                  | N |\n",
    "| Transcriptomics | TPM                  | N |\n",
    "| Transcriptomics | TMM                  | N |\n",
    "| Transcriptomics | RPKM                 | N |\n",
    "| Transcriptomics | UQ                   | N |\n",
    "| Transcriptomics | CTF                  | N |\n",
    "| Transcriptomics | CUF                  | N |\n",
    "| Transcriptomics | Others (UPDATE THIS) | N |\n",
    "\n",
    "\n",
    "## Data transformation\n",
    "\n",
    "\n",
    "| Data type      | Transformation                             | Reference |\n",
    "|----------------|--------------------------------------------|-----------|\n",
    "| Transcriptomics | asinh                                     | [Johnson and Krishnan 2022](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02568-9) |\n",
    "| Transcriptomics | Variance stabilizing transformation (VST) | DESeq2; used in [Priya et al 2022](https://www.nature.com/articles/s41564-022-01121-z) |\n",
    "\n",
    "\n",
    "## Filtering low variance (expression or OTUs)\n",
    "\n",
    "\n",
    "[Priya et al 2022](https://www.nature.com/articles/s41564-022-01121-z) used 25% quantile as cutoff for gene expression analysis.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Obtaining the two RNAseq matrices with estimated counts (day and night)\n",
    "\n",
    "File `0_kremling_expression_key.txt` from the `0_scripts` folder on the publication-associated [FigShare record](https://doi.org/10.6084/m9.figshare.5886769.v2) has association of identifiers between samples. The NCBI SRA records have associations between the unique identifiers in each dataset (RNAseq or 16S) and the associations with the SRA unique identifiers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "484 sample pairs found.\n",
      "2 samples without 16S data.\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "kremling_expression_key = '/home/santosrac/Repositories/maize_microbiome_transcriptomics/correlations_rnaseq_metataxonomics/0_kremling_expression_key.txt'\n",
    "sra_run_table_16s = '/home/santosrac/Repositories/maize_microbiome_transcriptomics/16S_wallace2018/SraRunInfo_Wallace_etal_2018.csv'\n",
    "sra_run_table_rnaseq = '/home/santosrac/Repositories/maize_microbiome_transcriptomics/rnaseq_kremling2018/run_info/SraRunInfo_Kremling_etal_2018.csv'\n",
    "\n",
    "# Dictionary for storing information about the samples (SRA identifier, day, day period, genotype) and associated runs (16S and RNAseq)\n",
    "dict_wallace_kremling_2018 = {}\n",
    "\n",
    "# Dictionary that will store association between Wallace and Kremling identifiers according to the key file\n",
    "kremling_expression_key_dict = {}\n",
    "\n",
    "# Counter to keep track of the number of samples with both 16S and RNAseq data\n",
    "rnaseq_samples_with_16s = 0\n",
    "\n",
    "# Counter to keep track of the number of samples without 16S data\n",
    "no_16s = 0\n",
    "\n",
    "# Dictionary to associate all 16S and RNAseq run identifiers with their sample identifier\n",
    "run2my_sample_id = {}\n",
    "\n",
    "# Reading the key file and creating a dictionary with the identifiers\n",
    "with open(kremling_expression_key, 'r') as file:\n",
    "\n",
    "    _ = file.readline()\n",
    "\n",
    "    for line in file:\n",
    "        fields = line.strip().split('\\t')\n",
    "        \n",
    "        kremling_identifier = fields[0]\n",
    "        wallace_identifier = fields[1]\n",
    "\n",
    "        # Adding both identifiers (Wallace and Kremling) to the dictionary\n",
    "        kremling_expression_key_dict[kremling_identifier] = wallace_identifier\n",
    "        kremling_expression_key_dict[wallace_identifier] = kremling_identifier\n",
    "\n",
    "# Reading the RNAseq SraRunInfo file and storing information about the samples (SRA identifier, day, day period, genotype)\n",
    "with open(sra_run_table_rnaseq, 'r') as file:\n",
    "\n",
    "    _ = file.readline()\n",
    "\n",
    "    for line in file:\n",
    "        fields = line.strip().split(',')\n",
    "        fields2 = fields[11].split('_')\n",
    "        rnaseq_run_id = fields[0]\n",
    "        sample_id = fields2[1]\n",
    "        rnaseq_genotype = fields2[2]\n",
    "        day = ''\n",
    "        match = re.search(r'\\d+', sample_id)\n",
    "        unmatched_parts = re.split(r'\\d+', sample_id)\n",
    "        day_period = unmatched_parts[0]\n",
    "        if match:\n",
    "            day = int(match.group())\n",
    "        if sample_id.startswith('LMA') and rnaseq_genotype != '#N/A':\n",
    "            dict_wallace_kremling_2018[fields[11]] = {'run_accession_16s': '',\n",
    "                                    'run_accession_rnaseq': rnaseq_run_id,\n",
    "                                    'day': day,\n",
    "                                    'day_period': day_period,\n",
    "                                    'genotype_16s': '',\n",
    "                                    'genotype_rnaseq': rnaseq_genotype}\n",
    "\n",
    "# Reading the 16S SraRunInfo file and storing information about the samples (SRA identifier, day, day period, genotype)\n",
    "with open(sra_run_table_16s, 'r') as file:\n",
    "\n",
    "    _ = file.readline()\n",
    "\n",
    "    for line in file:\n",
    "        fields = line.strip().split(',')\n",
    "        fields2 = fields[11].split('.')\n",
    "        metataxonomics_run_id = fields[0]\n",
    "        day = int(fields2[1])\n",
    "        day_period = fields2[0]\n",
    "        for key, value in kremling_expression_key_dict.items():\n",
    "            if value == fields[11]:\n",
    "                if dict_wallace_kremling_2018[key]['day'] != day:\n",
    "                    print('Big problem!')\n",
    "                    print(day, dict_wallace_kremling_2018[key]['day'])\n",
    "                    print(dict_wallace_kremling_2018[key])\n",
    "                    print(value, fields[11], key)\n",
    "                    exit(1)\n",
    "                if dict_wallace_kremling_2018[key]['day_period'] != day_period:\n",
    "                    # There is a single sample that has a different day period in the 16S data\n",
    "                    if key == '10343927_LMAN8_CML505_CAACAG':\n",
    "                        continue\n",
    "                    else:\n",
    "                        print(day_period, dict_wallace_kremling_2018[key]['day_period'])\n",
    "                        print(dict_wallace_kremling_2018[key])\n",
    "                        print(value, fields[11], key)\n",
    "                        exit(1)\n",
    "                dict_wallace_kremling_2018[key]['run_accession_16s'] = metataxonomics_run_id\n",
    "                rnaseq_samples_with_16s+=1\n",
    "\n",
    "# Checking for samples without 16S data\n",
    "for key, value in dict_wallace_kremling_2018.items():\n",
    "    if value['run_accession_16s'] == '':\n",
    "        no_16s+=1\n",
    "\n",
    "# Creating a dictionary to associate all 16S and RNAseq run identifiers with their sample identifier (sample identifier in SraRun table of RNAseq)\n",
    "for key in dict_wallace_kremling_2018:\n",
    "    if dict_wallace_kremling_2018[key]['run_accession_rnaseq']:\n",
    "        run2my_sample_id[dict_wallace_kremling_2018[key]['run_accession_rnaseq']] = key\n",
    "    if dict_wallace_kremling_2018[key]['run_accession_16s']:\n",
    "        run2my_sample_id[dict_wallace_kremling_2018[key]['run_accession_16s']] = key\n",
    "\n",
    "print(f'{rnaseq_samples_with_16s} sample pairs found.')\n",
    "print(f'{no_16s} samples without 16S data.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The file with the estimated RNAseq counts generated with Salmon can be found in the [wiki describing analyses for IPBGG retret 2024](https://github.com/SantosRAC/maize_microbiome_transcriptomics/wiki/IPBGG-2024-retreat) (`Zma2_counts_matrix.tar.gz`).\n",
    "\n",
    "Importing this matrix as a pandas dataframe and renaming the columns according to the sample names. It will be important for later filtering based on the day period."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 39096 genes/transcripts and 486 samples in the expression matrix.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>10343927_LMAD26_CI21E_AAGTGG</th>\n",
       "      <th>10343264_LMAN26_CI21E_ATGAAC</th>\n",
       "      <th>10343927_LMAN8_B73_CACACT</th>\n",
       "      <th>10343264_LMAN26_B64_ACCAGT</th>\n",
       "      <th>10343262_LMAN8_B109_TGCTAT</th>\n",
       "      <th>10343262_LMAN8_B14A_CTCTCG</th>\n",
       "      <th>10343262_LMAN8_B57_CCTAAG</th>\n",
       "      <th>10343927_LMAD26_B77_TAATCG</th>\n",
       "      <th>10343262_LMAN8_B79_GCAGCC</th>\n",
       "      <th>10343927_LMAN8_CI187-2_GACGAT</th>\n",
       "      <th>...</th>\n",
       "      <th>10344826_LMAN8_I29_ACGTCT</th>\n",
       "      <th>10344823_LMAD8_IA2132_ACACGC</th>\n",
       "      <th>10343264_LMAD26_CML91_AACGCC</th>\n",
       "      <th>10344827_LMAN26_CML91_AATCCG</th>\n",
       "      <th>10344827_LMAN26_Ki21_AAGACA</th>\n",
       "      <th>10343927_LMAD26_Ki21_ACGTCT</th>\n",
       "      <th>10344826_LMAD8_E2558W_CGCAAC</th>\n",
       "      <th>10343927_LMAN8_E2558W_GAACCT</th>\n",
       "      <th>10344826_LMAD8_IDS69_CAGGAC</th>\n",
       "      <th>10343927_LMAN8_IDS69_ACATTA</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Zm00001eb371370_T002</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zm00001eb371350_T001</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zm00001eb371330_T001</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zm00001eb371310_T001</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zm00001eb371280_T001</th>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>37</td>\n",
       "      <td>21</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 486 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      10343927_LMAD26_CI21E_AAGTGG  \\\n",
       "Name                                                 \n",
       "Zm00001eb371370_T002                             2   \n",
       "Zm00001eb371350_T001                             0   \n",
       "Zm00001eb371330_T001                             0   \n",
       "Zm00001eb371310_T001                             0   \n",
       "Zm00001eb371280_T001                             3   \n",
       "\n",
       "                      10343264_LMAN26_CI21E_ATGAAC  10343927_LMAN8_B73_CACACT  \\\n",
       "Name                                                                            \n",
       "Zm00001eb371370_T002                             0                          8   \n",
       "Zm00001eb371350_T001                             0                          0   \n",
       "Zm00001eb371330_T001                             0                          0   \n",
       "Zm00001eb371310_T001                             0                          0   \n",
       "Zm00001eb371280_T001                             6                          2   \n",
       "\n",
       "                      10343264_LMAN26_B64_ACCAGT  10343262_LMAN8_B109_TGCTAT  \\\n",
       "Name                                                                           \n",
       "Zm00001eb371370_T002                           0                           0   \n",
       "Zm00001eb371350_T001                           0                           0   \n",
       "Zm00001eb371330_T001                           0                           0   \n",
       "Zm00001eb371310_T001                           0                           0   \n",
       "Zm00001eb371280_T001                           0                          11   \n",
       "\n",
       "                      10343262_LMAN8_B14A_CTCTCG  10343262_LMAN8_B57_CCTAAG  \\\n",
       "Name                                                                          \n",
       "Zm00001eb371370_T002                           2                          1   \n",
       "Zm00001eb371350_T001                           0                          0   \n",
       "Zm00001eb371330_T001                           0                          0   \n",
       "Zm00001eb371310_T001                           0                          0   \n",
       "Zm00001eb371280_T001                           2                         11   \n",
       "\n",
       "                      10343927_LMAD26_B77_TAATCG  10343262_LMAN8_B79_GCAGCC  \\\n",
       "Name                                                                          \n",
       "Zm00001eb371370_T002                           3                          1   \n",
       "Zm00001eb371350_T001                           0                          0   \n",
       "Zm00001eb371330_T001                           0                          0   \n",
       "Zm00001eb371310_T001                           0                          0   \n",
       "Zm00001eb371280_T001                           0                          3   \n",
       "\n",
       "                      10343927_LMAN8_CI187-2_GACGAT  ...  \\\n",
       "Name                                                 ...   \n",
       "Zm00001eb371370_T002                              1  ...   \n",
       "Zm00001eb371350_T001                              0  ...   \n",
       "Zm00001eb371330_T001                              0  ...   \n",
       "Zm00001eb371310_T001                              0  ...   \n",
       "Zm00001eb371280_T001                              3  ...   \n",
       "\n",
       "                      10344826_LMAN8_I29_ACGTCT  10344823_LMAD8_IA2132_ACACGC  \\\n",
       "Name                                                                            \n",
       "Zm00001eb371370_T002                          0                             3   \n",
       "Zm00001eb371350_T001                          0                             0   \n",
       "Zm00001eb371330_T001                          0                             0   \n",
       "Zm00001eb371310_T001                          0                             0   \n",
       "Zm00001eb371280_T001                          4                             1   \n",
       "\n",
       "                      10343264_LMAD26_CML91_AACGCC  \\\n",
       "Name                                                 \n",
       "Zm00001eb371370_T002                             0   \n",
       "Zm00001eb371350_T001                             0   \n",
       "Zm00001eb371330_T001                             0   \n",
       "Zm00001eb371310_T001                             0   \n",
       "Zm00001eb371280_T001                             1   \n",
       "\n",
       "                      10344827_LMAN26_CML91_AATCCG  \\\n",
       "Name                                                 \n",
       "Zm00001eb371370_T002                             0   \n",
       "Zm00001eb371350_T001                             0   \n",
       "Zm00001eb371330_T001                             0   \n",
       "Zm00001eb371310_T001                             0   \n",
       "Zm00001eb371280_T001                             0   \n",
       "\n",
       "                      10344827_LMAN26_Ki21_AAGACA  \\\n",
       "Name                                                \n",
       "Zm00001eb371370_T002                            0   \n",
       "Zm00001eb371350_T001                            0   \n",
       "Zm00001eb371330_T001                            0   \n",
       "Zm00001eb371310_T001                            0   \n",
       "Zm00001eb371280_T001                           11   \n",
       "\n",
       "                      10343927_LMAD26_Ki21_ACGTCT  \\\n",
       "Name                                                \n",
       "Zm00001eb371370_T002                            9   \n",
       "Zm00001eb371350_T001                            0   \n",
       "Zm00001eb371330_T001                            0   \n",
       "Zm00001eb371310_T001                            0   \n",
       "Zm00001eb371280_T001                           37   \n",
       "\n",
       "                      10344826_LMAD8_E2558W_CGCAAC  \\\n",
       "Name                                                 \n",
       "Zm00001eb371370_T002                            11   \n",
       "Zm00001eb371350_T001                             0   \n",
       "Zm00001eb371330_T001                             0   \n",
       "Zm00001eb371310_T001                             0   \n",
       "Zm00001eb371280_T001                            21   \n",
       "\n",
       "                      10343927_LMAN8_E2558W_GAACCT  \\\n",
       "Name                                                 \n",
       "Zm00001eb371370_T002                             0   \n",
       "Zm00001eb371350_T001                             0   \n",
       "Zm00001eb371330_T001                             0   \n",
       "Zm00001eb371310_T001                             0   \n",
       "Zm00001eb371280_T001                             9   \n",
       "\n",
       "                      10344826_LMAD8_IDS69_CAGGAC  10343927_LMAN8_IDS69_ACATTA  \n",
       "Name                                                                            \n",
       "Zm00001eb371370_T002                            7                            0  \n",
       "Zm00001eb371350_T001                            0                            0  \n",
       "Zm00001eb371330_T001                            0                            0  \n",
       "Zm00001eb371310_T001                            0                            0  \n",
       "Zm00001eb371280_T001                            3                            4  \n",
       "\n",
       "[5 rows x 486 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Importing expression data from Kremling et al. 2018 (estimated counts matrix on Maize v5 using Salmon after cleaning with TrimGalore)\n",
    "# Values in the matrix are rounded to the nearest integer to avoid issues later when normalizing data\n",
    "kremling_expression_v5 = pd.read_csv('/home/santosrac/Projects/UGA_RACS/Correlations_RNAseq_16S/Zma2_counts_matrix.txt', sep='\\t')\n",
    "kremling_expression_v5.set_index('Name', inplace=True)\n",
    "kremling_expression_v5 = kremling_expression_v5.round(0).astype(int)\n",
    "\n",
    "# Renaming columns to match the sample identifiers in the dictionary\n",
    "kremling_expression_v5 = kremling_expression_v5.rename(columns=run2my_sample_id)\n",
    "kremling_expression_v5.columns = [str(x) for x in kremling_expression_v5.columns]\n",
    "print(f'There are {kremling_expression_v5.shape[0]} genes/transcripts and {kremling_expression_v5.shape[1]} samples in the expression matrix.')\n",
    "kremling_expression_v5.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creating the association dictionary between RNAseq identifiers and the identifiers given to samples in the OTU table comprising merged day and night counts.\n",
    "\n",
    "Then, creating the dataframe with separate RNAseq data for day and night samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 211 day samples\n",
      "There are 275 night samples\n",
      "RNAseq sample 10343264_LMAN26_B73_CCTGCT is associated with duplicated day identifier\n",
      "RNAseq sample 10343262_LMAN8_B73_GCGCTG is associated with duplicated day identifier\n",
      "RNAseq sample 10343927_LMAN8_B73_GCGCTG is associated with duplicated day identifier\n",
      "RNAseq sample 10344827_LMAN26_B73_CCTGCT is associated with duplicated day identifier\n",
      "Deleting 10343264_LMAN26_B73_CCTGCT (14A0029_26)\n",
      "Deleting 10343262_LMAN8_B73_GCGCTG (14A0113_8)\n",
      "Deleting 10343927_LMAN8_B73_GCGCTG (14A0113_8)\n",
      "Deleting 10344827_LMAN26_B73_CCTGCT (14A0029_26)\n",
      "RNAseq sample 10343927_LMAD8_CML505_AACGCC is associated with duplicated day identifier\n",
      "RNAseq sample 10343927_LMAN8_CML505_CAACAG is associated with duplicated day identifier\n",
      "Deleting 10343927_LMAD8_CML505_AACGCC (14A0570_8)\n",
      "Deleting 10343927_LMAN8_CML505_CAACAG (14A0570_8)\n",
      "RNAseq Day dataframe has 209 samples\n",
      "RNAseq Night dataframe has 271 samples\n"
     ]
    }
   ],
   "source": [
    "# Dictionaries with the association between the Kremling identifiers and the merged day and night identifiers\n",
    "kremling2_day_samples_dict = {}\n",
    "kremling2_night_samples_dict = {}\n",
    "\n",
    "for key in dict_wallace_kremling_2018.keys():\n",
    "    fields = kremling_expression_key_dict[key].split('.')\n",
    "    if fields[0] == 'LMAD':\n",
    "        kremling2_day_samples_dict[key] = fields[2]+'_'+str(fields[1])\n",
    "    elif fields[0] == 'LMAN':\n",
    "        kremling2_night_samples_dict[key] = fields[2]+'_'+str(fields[1])\n",
    "\n",
    "print(f'There are {len(kremling2_day_samples_dict)} day samples')\n",
    "print(f'There are {len(kremling2_night_samples_dict)} night samples')\n",
    "\n",
    "# Checking for duplicate identifiers and deleting duplicated records\n",
    "duplicated_day = []\n",
    "duplicated_night = []\n",
    "\n",
    "for key in kremling2_night_samples_dict.keys():\n",
    "    if kremling2_night_samples_dict[key] in [x for x in list(kremling2_night_samples_dict.values()) if list(kremling2_night_samples_dict.values()).count(x) > 1]:\n",
    "        print(f'RNAseq sample {key} is associated with duplicated day identifier')\n",
    "        duplicated_night.append(key)\n",
    "\n",
    "for key in duplicated_night:\n",
    "    print(f'Deleting {key} ({kremling2_night_samples_dict[key]})')\n",
    "    del kremling2_night_samples_dict[key]\n",
    "\n",
    "for key in kremling2_day_samples_dict.keys():\n",
    "    if kremling2_day_samples_dict[key] in [x for x in list(kremling2_day_samples_dict.values()) if list(kremling2_day_samples_dict.values()).count(x) > 1]:\n",
    "        print(f'RNAseq sample {key} is associated with duplicated day identifier')\n",
    "        duplicated_day.append(key)\n",
    "\n",
    "for key in duplicated_day:\n",
    "    print(f'Deleting {key} ({kremling2_day_samples_dict[key]})')\n",
    "    del kremling2_day_samples_dict[key]\n",
    "\n",
    "# Creating a dataframe with RNAseq day samples\n",
    "kremling_expression_v5_day = kremling_expression_v5.filter(items=list(kremling2_day_samples_dict.keys()))\n",
    "print(f'RNAseq Day dataframe has {kremling_expression_v5_day.shape[1]} samples')\n",
    "# Creating a dataframe with RNAseq night samples\n",
    "kremling_expression_v5_night = kremling_expression_v5.filter(items=list(kremling2_night_samples_dict.keys()))\n",
    "print(f'RNAseq Night dataframe has {kremling_expression_v5_night.shape[1]} samples')\n",
    "\n",
    "# Renaming dataframes with RNAseq samples\n",
    "kremling_expression_v5_day = kremling_expression_v5_day.rename(columns=kremling2_day_samples_dict)\n",
    "kremling_expression_v5_night = kremling_expression_v5_night.rename(columns=kremling2_night_samples_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Estimated counts datasets to be used in downstream analyses (RNAseq), before pairing with OTU matrix:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimensions of the Day matrix: (39096, 209)\n",
      "Dimensions of the Night matrix: (39096, 271)\n"
     ]
    }
   ],
   "source": [
    "print(f'Dimensions of the Day matrix: {kremling_expression_v5_day.shape}')\n",
    "print(f'Dimensions of the Night matrix: {kremling_expression_v5_night.shape}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Obtaining OTU data with merged day and night samples\n",
    "\n",
    "OTU counts for day and night samples were merged in a separate notebook and imported. Values were rounded and stored as integer to avoid problems in future normalization/transformation methods."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>14A0247_8</th>\n",
       "      <th>14A0051_8</th>\n",
       "      <th>14A0381_26</th>\n",
       "      <th>14A0533_26</th>\n",
       "      <th>14A0281_8</th>\n",
       "      <th>14A0295_8</th>\n",
       "      <th>14A0169_26</th>\n",
       "      <th>14A0069_8</th>\n",
       "      <th>14A0497_26</th>\n",
       "      <th>14A0023_8</th>\n",
       "      <th>...</th>\n",
       "      <th>14A0345_8</th>\n",
       "      <th>14A0267_8</th>\n",
       "      <th>14A0009_8</th>\n",
       "      <th>14A0007_8</th>\n",
       "      <th>14A0093_26</th>\n",
       "      <th>14A0137_26</th>\n",
       "      <th>14A0265_8</th>\n",
       "      <th>14A0155_26</th>\n",
       "      <th>14A0167_26</th>\n",
       "      <th>14A0481_26</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4479944</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995900</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1124709</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541139</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>533625</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>40</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 249 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         14A0247_8  14A0051_8  14A0381_26  14A0533_26  14A0281_8  14A0295_8  \\\n",
       "OTU ID                                                                        \n",
       "4479944          1          2           3           1          0          0   \n",
       "995900           0          0           0           0          5          8   \n",
       "1124709          0          0           0           0          0          0   \n",
       "541139           0          0           0           0          0          0   \n",
       "533625           0          1           2           0          0          0   \n",
       "\n",
       "         14A0169_26  14A0069_8  14A0497_26  14A0023_8  ...  14A0345_8  \\\n",
       "OTU ID                                                 ...              \n",
       "4479944           0          0           0          0  ...          0   \n",
       "995900            6          1           1          1  ...          0   \n",
       "1124709           0          0           0          0  ...          0   \n",
       "541139            0          0           0          0  ...          0   \n",
       "533625            0          0          40          1  ...          0   \n",
       "\n",
       "         14A0267_8  14A0009_8  14A0007_8  14A0093_26  14A0137_26  14A0265_8  \\\n",
       "OTU ID                                                                        \n",
       "4479944          0          0          0           0           0          0   \n",
       "995900           0          0          0           0           0          0   \n",
       "1124709          0          0          0           0           0          0   \n",
       "541139           0          0          0           0           0          0   \n",
       "533625           0          0          1           0           0          0   \n",
       "\n",
       "         14A0155_26  14A0167_26  14A0481_26  \n",
       "OTU ID                                       \n",
       "4479944           0           0           0  \n",
       "995900            0           0           0  \n",
       "1124709           0           0           0  \n",
       "541139            0           0           0  \n",
       "533625            2           0           2  \n",
       "\n",
       "[5 rows x 249 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "otu_table_merged_d_n = pd.read_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/16S_wallace2018/combine_day_night_samples/summed_day_night_otu_counts.tsv',\n",
    "                                   sep='\\t', index_col='OTU ID')\n",
    "otu_table_merged_d_n = otu_table_merged_d_n.round(0).astype(int)\n",
    "\n",
    "otu_table_merged_d_n.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Obtaining pairs with both 16S and RNAseq data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimensios of the Day OTU table: (9057, 176)\n",
      "Dimensios of the Night OTU table: (9057, 228)\n"
     ]
    }
   ],
   "source": [
    "otu_table_merged_day = otu_table_merged_d_n.filter(items=kremling_expression_v5_day.columns)\n",
    "print(f'Dimensios of the Day OTU table: {otu_table_merged_day.shape}')\n",
    "otu_table_merged_night = otu_table_merged_d_n.filter(items=kremling_expression_v5_night.columns)\n",
    "print(f'Dimensios of the Night OTU table: {otu_table_merged_night.shape}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimensios of the Day expression table: (39096, 176)\n",
      "Dimensios of the Night expression table: (39096, 228)\n"
     ]
    }
   ],
   "source": [
    "kremling_expression_v5_day = kremling_expression_v5_day.filter(items=otu_table_merged_day.columns)\n",
    "print(f'Dimensios of the Day expression table: {kremling_expression_v5_day.shape}')\n",
    "kremling_expression_v5_night = kremling_expression_v5_night.filter(items=otu_table_merged_night.columns)\n",
    "print(f'Dimensios of the Night expression table: {kremling_expression_v5_night.shape}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exporting matrices to files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "otu_table_merged_day.to_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/correlations_rnaseq_metataxonomics/otu_table_merged_day.tsv', sep='\\t')\n",
    "otu_table_merged_night.to_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/correlations_rnaseq_metataxonomics/otu_table_merged_night.tsv', sep='\\t')\n",
    "kremling_expression_v5_day.to_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/correlations_rnaseq_metataxonomics/kremling_expression_v5_day.tsv', sep='\\t')\n",
    "kremling_expression_v5_night.to_csv('/home/santosrac/Repositories/maize_microbiome_transcriptomics/correlations_rnaseq_metataxonomics/kremling_expression_v5_night.tsv', sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
